src/lib/comms/sol-connman-impl-connman.c

```diff
@@ -0,0 +1,910 @@
+    SOL_INT_CHECK(r, < 0, r);
+
+    do {
+        r = sd_bus_message_enter_container(m, SD_BUS_TYPE_DICT_ENTRY, "sv");
+        SOL_INT_CHECK_GOTO(r, < 1, end);
+
+        r = sd_bus_message_read_basic(m, SD_BUS_TYPE_STRING, &str);
+        SOL_INT_CHECK(r, < 0, r);
+        if (strcmp(str, "Address") == 0) {
+            char *address;
+
+            r = sd_bus_message_enter_container(m, SD_BUS_TYPE_VARIANT, "s");
+            SOL_INT_CHECK(r, < 0, r);
+
+            r = sd_bus_message_read_basic(m, SD_BUS_TYPE_STRING, &address);
+            SOL_INT_CHECK(r, < 0, r);
+
+            sol_network_link_addr_from_str(link_addr, address);
+
+            r = sd_bus_message_exit_container(m);
+            SOL_INT_CHECK(r, < 0, r);
+        } else {
+            r = sd_bus_message_skip(m, "v");
+            SOL_INT_CHECK(r, < 0, r);
+        }
+
+        r = sd_bus_message_exit_container(m);
+        SOL_INT_CHECK(r, < 0, r);
+    } while (1);
+
+end:
+    if (r == 0)
+        r = sd_bus_message_exit_container(m);
+
+    return r;
+}
+
+static void
+get_service_ipv4(sd_bus_message *m, struct sol_connman_service *service)
+{
+    if (!service)
+        return;
+
+    if (!service->link_addr) {
+        service->link_addr = malloc(sizeof(*service->link_addr));
+        if (!service->link_addr)
+            return;
+        service->link_addr->family = SOL_NETWORK_FAMILY_INET;
+    }
+    get_service_ip(m, service->link_addr);
+}
+
+static void
+get_service_ipv6(sd_bus_message *m, struct sol_connman_service *service)
+{
+    if (!service)
+        return;
+
+    if (!service->link_addr6) {
+        service->link_addr6 = malloc(sizeof(*service->link_addr6));
+        if (!service->link_addr6)
+            return;
+        service->link_addr6->family = SOL_NETWORK_FAMILY_INET6;
+    }
+    get_service_ip(m, service->link_addr6);
+}
+
+static void
+remove_services(const char *path)
+{
+    struct sol_connman_service *service;
+    uint16_t i;
+
+    if (!path)
+        return;
+
+    SOL_VECTOR_FOREACH_IDX (&_ctx.service_vector, service, i) {
+        if (strcmp(service->path, path) == 0) {
+            service->state = SOL_CONNMAN_SERVICE_STATE_REMOVE;
+            call_monitor_callback(service);
+            _free_connman_service(service);
+            sol_vector_del(&_ctx.service_vector, i);
+            break;
+        }
+    }
+}
+
+static int
+get_services_properties(sd_bus_message *m, const char *path)
+{
+    int r;
+    struct sol_connman_service *service;
+    uint16_t i;
+    bool is_exist = false;
+
+    SOL_VECTOR_FOREACH_IDX (&_ctx.service_vector, service, i) {
+        if (strcmp(service->path, path) == 0) {
+            is_exist = true;
+            break;
+        }
+    }
+
+    if (is_exist == false) {
+        service = sol_vector_append(&_ctx.service_vector);
+        SOL_NULL_CHECK(service, -EINVAL);
+
+        service->path = strdup(path);
+        SOL_NULL_CHECK(service->path, -ENOMEM);
```

Comment:
here you leak the element in the array, you need to `SOL_NULL_CHECK_GOTO()` and `sol_vector_del_last()`